src/compiler/checker.ts

```diff
@@ -3149,20 +3174,23 @@ namespace ts {
             }
         }
 
-        function getPropertiesOfUnionType(type: UnionType): Symbol[] {
-            let result: Symbol[] = [];
-            forEach(getPropertiesOfType(type.types[0]), prop => {
-                let unionProp = getPropertyOfUnionType(type, prop.name);
-                if (unionProp) {
-                    result.push(unionProp);
+        function getPropertiesOfUnionOrIntersectionType(type: UnionOrIntersectionType): Symbol[] {
+            for (let current of type.types) {
+                for (let prop of getPropertiesOfType(current)) {
+                    getPropertyOfUnionOrIntersectionType(type, prop.name);
                 }
-            });
-            return result;
+                // The properties of a union type are those that are present in all constituent types, so
+                // we only need to check the properties of the first type
+                if (type.flags & TypeFlags.Union) {
+                    break;
+                }
+            }
+            return type.resolvedProperties ? symbolsToArray(type.resolvedProperties) : emptyArray;
         }
 
         function getPropertiesOfType(type: Type): Symbol[] {
             type = getApparentType(type);
-            return type.flags & TypeFlags.Union ? getPropertiesOfUnionType(<UnionType>type) : getPropertiesOfObjectType(type);
+            return type.flags & TypeFlags.UnionOrIntersection ? getPropertiesOfUnionOrIntersectionType(<UnionType>type) : getPropertiesOfObjectType(type);

```

Comment:
wrap this line.  it is too long for github.  I recommend the form:

```
return type.flags & TypeFlags.UnionOrIntersection
    ? getPropertiesOfUnionOrIntersectionType(<UnionOrIntersectionType>type)
    : getPropertiesOfObjectType(type);
```

This also has the benefit of being able to easily see the difference in the branches.